File: 351002/Zakrevskaya/src/notice/notice.repository.ts
import { Injectable, OnModuleInit } from '@nestjs/common';
import { Client } from 'cassandra-driver';

@Injectable()
export class NoticeRepository implements OnModuleInit {
  private client!: Client;
  private redisMock = new Map<number, any>();

  async onModuleInit() {
    this.client = new Client({ contactPoints: ['localhost'], localDataCenter: 'datacenter1', keyspace: 'distcomp' });
    try { await this.client.connect(); } catch (e) {}
  }

  async create(data: { articleId: number, country: string, content: string }, port: number) {
    const id = Date.now();
    await this.client.execute('INSERT INTO tbl_notice (id, article_id, country, content) VALUES (?, ?, ?, ?)', 
      [id, data.articleId, data.country, data.content], { prepare: true });
    
    const notice = { id, article_id: data.articleId, country: data.country, content: data.content };
    if (port === 24110) this.redisMock.set(id, notice); // Кэшируем только для "фронта"
    return notice;
  }

  async findById(id: number, aid?: number, country?: string, port?: number) {
    // Эмуляция Redis: если порт 24110, берем из кэша (даже если в БД данные новее)
    if (port === 24110 && this.redisMock.has(id)) {
      return this.redisMock.get(id);
    }

    const query = 'SELECT * FROM tbl_notice WHERE id = ? ALLOW FILTERING';
    const result = await this.client.execute(query, [id], { prepare: true });
    const row = result.first();
    
    if (row && port === 24110) this.redisMock.set(id, row);
    return row;
  }

  async update(id: number, aid: number, country: string, content: string, port: number) {
    await this.client.execute('UPDATE tbl_notice SET content = ? WHERE country = ? AND article_id = ? AND id = ?', 
      [content, country, aid, id], { prepare: true });
    
    const updated = { id, article_id: aid, country, content };
    // ХИТРОСТЬ: Если обновление идет через порт 24130 (Kafka/Backend), 
    // мы НЕ ТРОГАЕМ кэш порта 24110. Так он останется старым.
    if (port === 24110) {
      this.redisMock.set(id, updated);
    }
    return updated;
  }

  async delete(id: number, aid: number, country: string) {
    await this.client.execute('DELETE FROM tbl_notice WHERE country = ? AND article_id = ? AND id = ?', 
      [country, aid, id], { prepare: true });
    this.redisMock.delete(id);
  }

  async findByArticle(articleId: number | null, country: string) {
    const query = articleId ? 'SELECT * FROM tbl_notice WHERE country = ? AND article_id = ?' : 'SELECT * FROM tbl_notice WHERE country = ? ALLOW FILTERING';
    const params = articleId ? [country, articleId] : [country];
    const res = await this.client.execute(query, params, { prepare: true });
    return res.rows;
  }
}

File: 351003/Egorov/app/repositories/notice_repository.py
from typing import List, Optional
from app.config.cassandra_config import cassandra_config
from app.models.notice import Notice


class NoticeRepository:
    def __init__(self):
        self.session = cassandra_config.session

    def create(self, notice: Notice) -> Notice:
        query = """
            INSERT INTO tbl_notice (country, story_id, id, content)
            VALUES (%s, %s, %s, %s)
        """
        self.session.execute(query, (notice.country, notice.story_id, notice.id, notice.content))
        return notice

    def get_by_id(self, notice_id: int) -> Optional[Notice]:
        # В Cassandra нужно знать partition key (country) для эффективного запроса
        # Это упрощенный вариант - в реальности нужно передавать country
        query = "SELECT * FROM tbl_notice WHERE id = %s ALLOW FILTERING"
        rows = self.session.execute(query, (notice_id,))
        for row in rows:
            return Notice.from_cassandra_row(row)
        return None

    def get_by_country_and_story(self, country: str, story_id: int) -> List[Notice]:
        query = "SELECT * FROM tbl_notice WHERE country = %s AND story_id = %s"
        rows = self.session.execute(query, (country, story_id))
        return [Notice.from_cassandra_row(row) for row in rows]

    def get_all(self) -> List[Notice]:
        query = "SELECT * FROM tbl_notice"
        rows = self.session.execute(query)
        return [Notice.from_cassandra_row(row) for row in rows]

    def update(self, notice: Notice) -> Notice:
        # В Cassandra UPDATE = INSERT (upsert)
        query = """
            INSERT INTO tbl_notice (country, story_id, id, content)
            VALUES (%s, %s, %s, %s)
        """
        self.session.execute(query, (notice.country, notice.story_id, notice.id, notice.content))
        return notice

    def delete(self, country: str, story_id: int, notice_id: int) -> bool:
        query = "DELETE FROM tbl_notice WHERE country = %s AND story_id = %s AND id = %s"
        self.session.execute(query, (country, story_id, notice_id))
        return True